### PR TITLE
Add types for `foundry.abstract.DatabaseBackend` and `ClientDatabaseBackend`

### PR DESCRIPTION
**Pull request by In3luki**

### CODE REVIEW

File: types/foundry/common/abstract/backend.d.ts [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        request: DatabaseBackendGetContext,
+        user: User
+    ): Promise<(DeepPartial<TDocument["_source"][]> & CompendiumIndexData) | TDocument[]>;
+
+    /**
+     * Get the parent Document (if any) associated with a request context.
+     * @param context    The requested operation
+     * @return           The parent Document, or null
+     */
+    private _getParent(
```
Comment (stwlam) [inline types/foundry/common/abstract/backend.d.ts:43]: ?

Comment (In3luki) [reply types/foundry/common/abstract/backend.d.ts:43]:
```js
  /**
   * Get the parent Document (if any) associated with a request context.
   * @param {object} context                The requested operation
   * @return {Promise<Document|null>}       The parent Document, or null
   * @private
   */
  async _getParent(context) {
    if ( !context.parent ) return null;
    if ( !(context.parent instanceof Document) ) {
      throw new Error("A parent Document provided to the database operation must be a Document instance");
    }
    return context.parent;
  }
  ```

Comment (stwlam) [reply types/foundry/common/abstract/backend.d.ts:43]: If it's truly private, it doesn't need to be there since it's not part of the actual interface.

Comment (In3luki) [reply types/foundry/common/abstract/backend.d.ts:43]: Ok, I wasn't sure what to do with the `@private` methods. I'll remove them when I'm home.

Comment (In3luki) [reply types/foundry/common/abstract/backend.d.ts]: They are gone

**Pull request merged**